File: java/2015/day05b/Program.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.regex.*;

public class Program {
    public static void usage() {
        System.err.println("usage: java -jar Program.jar <input file>");
        System.exit(1);
    }

    public static boolean prop1(String line) {
        Pattern p = Pattern.compile("(..).*\\1");
        Matcher m = p.matcher(line);
        return m.find();
    }

    public static boolean prop2(String line) {
        Pattern p = Pattern.compile("(.).\\1");
        Matcher m = p.matcher(line);
        return m.find();
    }

    public static int process(String filename) {
        int result = 0;
        try ( BufferedReader reader = new BufferedReader(new FileReader(filename))) {
            String line;
            while ( (line = reader.readLine()) != null ) {
                if ( prop1(line) && prop2(line) ) {
                    result++;
                }
            }
        } catch ( IOException e ) {
            System.err.println("Error reading file: " + e.getMessage());
            System.exit(1);
        }
        return result;
    }

    public static final void main(String[] args) {
        if ( args.length < 1 ) {
            usage();
        }
        String filename = args[0];
        int result = process(filename);
        System.out.println("result = " + result);
    }
}

File: scala/2015/day05b/Program.scala
import scala.io.Source
import scala.util.Using
import scala.sys.exit

object Program {
  def usage(): Unit = {
    Console.err.println("usage: scala Program <input file>")
    sys.exit(1)
  }

  def process(filename: String): Int = {
    def prop1(str: String) : Boolean = {
      "(..).*\\1".r.findFirstIn(str).isDefined
    }

    def prop2(str: String) : Boolean = {
      "(.).\\1".r.findFirstIn(str).isDefined
    }

    Using(Source.fromFile(filename)) { source =>
      source.getLines().count(s => prop1(s) && prop2(s))
    }.getOrElse(0)
  }

  def main(args: Array[String]): Unit = {
    args.toList match {
      case filename :: Nil =>
        val result = process(filename)
        println(s"result = $result")
      case _ =>
        usage()
    }
  }
}
